torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -324,46 +330,57 @@ void ProcessGroupAgent::enqueueSend(SendWork work) {
   // NB: this can be changed to use a native move capture when moved to C++14
   threadPool_.run(std::bind(
       [this](const SendWork& work) {
-        std::string serializedPayload =
-            wireSerialize(work.message_.payload(), work.message_.tensors());
-
-        std::vector<torch::Tensor> preamble = {torch::tensor(
-            {(int64_t)pg_->getRank(),
-             (int64_t)serializedPayload.length(),
-             (int64_t)work.message_.type(),
-             (int64_t)work.message_.id()},
-            {torch::kInt64})};
-
-        // ProcessGroup is not thread-safe when sending with the same tag, hence
-        // the lock
-        std::vector<std::shared_ptr<c10d::ProcessGroup::Work>> pendingSends;
-        const auto dst = work.to_.id_;
-        if (work.message_.isShutdown()) {
-          pendingSends.reserve(1);
-          {
-            std::lock_guard<std::mutex> guard(sendMutexes_[dst]);
-            pendingSends.emplace_back(
-                pg_->send(preamble, dst, dst /* channelTag */));
+        try {
+          std::string serializedPayload =
```

Comment:
Can we move the existing code block to a separate function like `handleSend` and then just have a try-catch block here around `handleSend`? That way it's easier to read the code and understand error handling.

Reply:
- This part of the change probably needs to be rebased due to recent merge conflicts, and is now a bit shorter than it was previously.

- Re: code review culture - it's reasonable to ask for refactoring, but sometimes, the refactoring is best left for a future changes, so that incremental improvements more readable. Not sure what's best in this particular case - does it make sense deferring, or not?

Reply:
I think if the refactoring is not too complex, it's best to just add it in the current diff since:
1) Cutting an issue to do it later adds some more overhead and it may not be done for a while if it's not too high priority
2) If there are already necessary changes that need to be made on a PR, adding the refactoring won't slow us down since the PR needs to be resubmitted to CI/rebased anyways

But you do raise a good point on the readability perspective, a PR should be concise and not try to do more than what it describes.

Reply:
> But you do raise a good point on the readability perspective, a PR should be concise and not try to do more than what it describes.

Thanks, I completely agree with this particular point.

I believe there probably are some things that could be changed about the team's code review culture to help encourage this, since the current culture encourages somewhat the opposite.